src/Aeson.purs

```diff
@@ -112,24 +163,153 @@ getField' decoder obj str =
     (lmap (AtKey str) <<< decoder)
     (FO.lookup str obj)
 
+-- | Returns an Aeson available under a sequence of keys in given Aeson.
+--   If not possible returns JsonDecodeError.
+getNestedAeson :: Aeson -> Array String -> Either JsonDecodeError Aeson
+getNestedAeson asn@(Aeson { numberIndex, patchedJson: AesonPatchedJson pjson }) keys =
+  note (UnexpectedValue $ toStringifiedNumbersJson asn) $
+    mkAeson <$> (foldM lookup pjson keys :: Maybe Json)
+  where
+  lookup :: Json -> String -> Maybe Json
+  lookup j lbl = caseJsonObject Nothing (FO.lookup lbl) j
+
+  mkAeson :: Json -> Aeson
+  mkAeson json = Aeson { numberIndex, patchedJson: AesonPatchedJson json }
+
 infix 7 getField as .:
 
--- TODO: add getFieldOptional if ever needed.
+type AesonCases a =
+  { caseNull :: Unit -> a
+  , caseBoolean :: Boolean -> a
+  , caseNumber :: String -> a
+  , caseString :: String -> a
+  , caseArray :: Array Aeson -> a
+  , caseObject :: Object Aeson -> a
+  }
 
-foreign import parseJsonExtractingIntegers
-  :: (forall (a :: Type) (b :: Type). a -> b -> Tuple a b)
-  -> String
-  -> String /\ NumberIndex
+caseAeson
+  :: forall a
+   . AesonCases a
+  -> Aeson
+  -> a
+caseAeson
+  { caseNull, caseBoolean, caseNumber, caseString, caseArray, caseObject }
+  (Aeson { numberIndex, patchedJson: AesonPatchedJson pJson }) = caseJson
+  caseNull
+  caseBoolean
+  (coerceNumber >>> unsafeIndex numberIndex >>> caseNumber)
+  caseString
+  (map mkAeson >>> caseArray)
+  (map mkAeson >>> caseObject)
+  pJson
+  where
+  mkAeson :: Json -> Aeson
+  mkAeson json = Aeson { patchedJson: AesonPatchedJson json, numberIndex }
+
+  -- will never get index out of bounds
+  unsafeIndex :: forall (x :: Type). Array x -> Int -> x
+  unsafeIndex arr ix = unsafePartial $ Array.unsafeIndex arr ix
+
+  -- will never encounter non int number
+  coerceNumber :: Number -> Int
+  coerceNumber = round
+
+constAesonCases :: forall (a :: Type). (forall x. x -> a) -> AesonCases a
```

Comment:
`(forall x. x -> a)` is equivalent to `a` from the users point of view. Why not as for just `a`?